File: cxx/traaxx/sources/level_count_for.cpp
#include <traaxx/level_count_for.hpp>

namespace traaxx
{
    template<typename IndexT>
    IndexT level_count_for(IndexT d_max)
    {
        if (d_max <= IndexT{ 1 })
        {
            return IndexT{ 1 };
        }
        auto k = IndexT{ 0 };
        auto reach = IndexT{ 1 };
        while (reach < d_max)
        {
            reach = static_cast<IndexT>(reach * IndexT{ 2 });
            ++k;
        }
        return static_cast<IndexT>(k + IndexT{ 1 });
    }

    template std::uint32_t level_count_for<std::uint32_t>(std::uint32_t);
}

File: go/std/level_count_for.go
package std

// LevelCountFor computes the number of ancestor-table levels (K+1) needed to
// reach any node at distance up to dMax via pointer doubling: ceil(log2(dMax))
// plus one level of margin. Mirrors traaxx::level_count_for
// (cxx/traaxx/sources/level_count_for.cpp).
func LevelCountFor(dMax Index) Index {
	if dMax <= 1 {
		return 1
	}
	k := Index(0)
	reach := Index(1)
	for reach < dMax {
		reach *= 2
		k++
	}
	return k + 1
}
